# ANN

The artificial neuron receives one or more inputs (representing dendrites) and sums them to produce an output. <br/>
It receives the input signal from the external world in the form of a pattern and image in the form of a vector. These inputs are then mathematically designated by the notations x(n) for every n number of inputs.

# Problem : Classification Problem
 Analyse the Conversion Rate

In [1]:
#Importing Libraries
import numpy as np
import pandas as pd
import tensorflow as tf


In [2]:
#version of tensorflow
tf.__version__

'2.1.0'

In [3]:
pwd

'C:\\Users\\User'

In [4]:
cd C:\Users\User\Desktop\Anindita Mitra\Deep Learning\original (1)\Machine Learning A-Z (Codes and Datasets)\Part 8 - Deep Learning\Section 39 - Artificial Neural Networks (ANN)\Python

C:\Users\User\Desktop\Anindita Mitra\Deep Learning\original (1)\Machine Learning A-Z (Codes and Datasets)\Part 8 - Deep Learning\Section 39 - Artificial Neural Networks (ANN)\Python


In [5]:
#importing Data Set
df = pd.read_csv("Email_DL.csv")


In [17]:
df.head(10)

RowNumber  Email_ID Customer_Name  Open_Rate  Click_Rate  Conversion_Rate
0          1  15634602      Hargrave        619          42                1
1          2  15647311          Hill        608          41                0
2          3  15619304          Onio        502          42                1
3          4  15701354          Boni        699          39                0
4          5  15737888      Mitchell        850          43                0
5          6  15574012           Chu        645          44                1
6          7  15592531      Bartlett        822          50                0
7          8  15656148        Obinna        376          29                1
8          9  15792365            He        501          44                0
9         10  15592389            H?        684          27                0

In [6]:
#defining X & Y
x = df.iloc[:,3:-1].values
y=df.iloc[:,-1].values

In [7]:
print(x)

[[619  42]
 [608  41]
 [502  42]
 ...
 [709  36]
 [772  42]
 [792  28]]


In [8]:
print(y)

[1 0 1 ... 1 1 0]


In [9]:
from sklearn.model_selection import train_test_split

In [10]:
x_train, x_test, y_train, y_test = train_test_split (x , y, test_size=0.2, random_state=0)

# Initializing ANN

In [11]:
ann = tf.keras.models.Sequential()

In [12]:
# Adding input layer and first hidden layer
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

In [13]:
# Adding second hidden layer
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

In [14]:
# Adding output layer
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Compling the ANN

In [15]:
ann.compile(optimizer = 'adam', loss = 'binary_crossentropy', metrics = ['accuracy'])

# Training the ANN

In [16]:
ann.fit(x_train, y_train, batch_size = 32, epochs=100)

Train on 8000 samples
Epoch 1/100
8000/8000 [==============================] - 3s 419us/sample - loss: 4.2801 - accuracy: 0.6961
Epoch 2/100
8000/8000 [==============================] - 1s 92us/sample - loss: 0.4968 - accuracy: 0.7931
Epoch 3/100
8000/8000 [==============================] - 1s 110us/sample - loss: 0.4942 - accuracy: 0.7878
Epoch 4/100
8000/8000 [==============================] - 1s 137us/sample - loss: 0.4978 - accuracy: 0.7862
Epoch 5/100
8000/8000 [==============================] - 1s 78us/sample - loss: 0.4984 - accuracy: 0.7915
Epoch 6/100
8000/8000 [==============================] - 1s 79us/sample - loss: 0.4921 - accuracy: 0.7879
Epoch 7/100
8000/8000 [==============================] - 1s 90us/sample - loss: 0.5110 - accuracy: 0.7875
Epoch 8/100
8000/8000 [==============================] - 1s 84us/sample - loss: 0.4981 - accuracy: 0.7876
Epoch 9/100
8000/8000 [==============================] - 1s 74us/sample - loss: 0.5091 - accuracy: 0.7849
Epoch 10/100
8000/800

# Prediction

If Open Rate is 727 and click rate is 450 : what should be the conversion rate?


In [21]:
print(ann.predict([[727,450]]))

[[1.]]


# Predicting Y_Pred

In [24]:
y_pred = ann.predict(x_test)

In [26]:
y_pred = (y_pred > 0.5)

# Confusion Matrix

In [27]:
#Confusion Matrix
# No of correct and incorrt prdection
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[1584   11]
 [ 399    6]]


1584 : correct prediction : no conversion (0) <br/>
6 : Correct Prediction : yes conversion (1)<br/>
11 : Incorrect Prediction : yes conversion (1)<br/>
399 : Incorrect Prediction : n0 conversion(0)<br/>


In [28]:
#.82 % is the corrt prediction
accuracy_score(y_test, y_pred) 

0.795